Add VM installation guide (#366)
osism/metalbox: INSTALL-VM.md

# Metalbox installation on an existing server

## Prerequisites

The server needs to be running Ubuntu 24.04 LTS. We recommend having
at least 8 vCPUs, 32 GB RAM and 100 GB SSD disk.

## Basic preparation

Execute these commands as some user on the server that has sudo
access at least via a password. In the command example below
this user is named `osism`, amend the commands accordingly if
you are using a different username.

```
cd
sudo apt update
sudo apt full-upgrade -y
sudo apt install -y git pipx python3-venv sshpass
git clone https://github.com/osism/metalbox/
sudo mv metalbox /opt/configuration
cd /opt/configuration/environments/manager/
echo "ansible_connection: local" >> host_vars/metalbox/vars.yml
./run.sh operator
sudo chown -R dragon:dragon /opt/configuration
```

> Note the this creates the `dragon` user and allows SSH login for

Reviewer: ```suggestion
Note that the `./run.sh operator` creates the `dragon` user and allows SSH login for
```

it with a well-known SSH key that is contained in this repo. If your
VM is reachable from the Internet via SSH, you should replace that
key with some locally-generated secure SSH key.

## Log in again as dragon

The previous steps created the `dragon` user account which
will be used to run the Metalbox service. Log in as that user
either directly or by executing `sudo -iu dragon` from your
existing account.

## Prepare the Metalbox installation

Prepare the environment:

```
cd
git clone https://github.com/osism/openstack-ironic-images
cd /opt/configuration/environments/manager/

> ./run.sh noop

Reviewer: a comment here would be good to explain why we run noop

> sed -i 's/\(INSTALL_ANSIBLE_ROLES=${INSTALL_ANSIBLE_ROLES:-\)true/\1false/' /opt/configuration/environments/manager/run.sh

Reviewer: a comment here would be good to explain what the goal here is and why

Author: deleted as it is actually not used after this point

> venv/bin/ansible-playbook ~/openstack-ironic-images/elements/metalbox/static/root/part1.yml

Reviewer: what does part1 do? a short comment here would be good

pipx install netbox-manager

> pipx install 'ansible-core>=2.19.0,<2.20.0'

Reviewer: it would be good to explain what is done here since pipx might not be a well known thing yet

/home/dragon/.local/bin/ansible-galaxy collection install netbox.netbox
```

> Fill the registry volume and start the registry:

Reviewer: i think this needs a bit more context of what "the registry volume" is, what it is good for and what it is filled with

```
cd
wget -O registry.tar.bz2 https://nbg1.your-objectstorage.com/osism/metalbox/registry.tar.bz2
docker run --rm -v registry:/volume -v /home/dragon:/import library/alpine:3 sh -c 'cd /volume && tar xjf /import/registry.tar.bz2'

> docker run -d -p 0.0.0.0:5001:5000 -v registry:/var/lib/registry --name registry --restart always library/registry:3

Reviewer: i think short comments here would help the reader to understand what those complicated commands are doing and why they are run

```

> Add a well-known IP address in order for some API operations to work.

Reviewer: what are "some API operations" ? and what is a "well-known IP"?

> Depending on how your network is configured, you may want to add

Reviewer: what is the deciding factor here if i should want this?

this configuration to your permanent configuration, e.g. by adding

> it to a file in `/etc/netplan`.

Reviewer: i think it would be good to show how the netplan file would look like

```
sudo ip link add metalbox type dummy
sudo ip addr add 192.168.42.10/32 dev metalbox
sudo ip link set up metalbox
```

## Fill the netbox directory

In the directory containing your netbox data (this can be either a
dedicated git repo or the `netbox` subdirectory of your cloudpod

> repository) run:

Reviewer: i assume i should have copied this somehow somewhere to the metalbox server? or is it enough to copy the exported archive? do i clone the repository?

```
netbox-manager export-archive
netbox-manager import-archive
```

This will copy the relevant data into `/opt/configuration/netbox/`.

> Deploy the netbox and fill it with data:

Reviewer: ```suggestion
Deploy the netbox and import the data from `/opt/configuration/netbox`:
```

```
deploy-netbox.sh

> netbox-manage.sh 

Reviewer: i think here we should mention the option to increase the import speed by using parallel import

Author: This didn't work in my test case, I would prefer stability over speed, we need a way to verify that a netbox repository can actually imported in parallel before we document this

```

Each of the above commands can take a couple of minutes to complete.

Configure the site for the metalbox deployment. Replace
`$metalbox` with the device name of your Metalbox

> server in the Netbox:

Reviewer: tbh, i have no idea what this means. what site are we talking about? why would i want to configure any site for the metalbox at this point?

Author: A single metalbox is responsible for a single site, but the netbox repository can contain multiple sites, therefore it is necessary to tell the metalbox which site it is responsible for. This is the same as step 7 in the general installation guide. I'll just copy the wording from there to avoid confusion.

```
get-netbox-config.sh $metalbox
```

> ## Finish the Metalbox installation

Reviewer: I think it would probably be good to split the installation of the metalbox and the generation of inventory, sonic configuration and maybe even the import/sync of data from netbox into ironic. To ease the understanding here, maybe we can reference something from here https://osism.tech/docs/concepts/metalbox. We can also extend the information there.

Deploy the manager and generate sonic configs:

```
deploy-manager.sh 
osism sync inventory
osism sync sonic
```

> You can find the generated configs in `/opt/httpd/data/sonic/`.

Reviewer: i think this needs more context what configs and what i should look at for what information

Author: your are right, I replaced this with a more generic comment
